### PR TITLE
disableInitialLoad option, register slot callback, and npm audit

### PR DESCRIPTION
**Pull request by scott-thrillist**
This PR allows you to enable DFP's disableInitialLoad option [[link](https://developers.google.com/doubleclick-gpt/reference#googletag.PubAdsService_disableInitialLoad)]. Also added a a callback prop that allows you to fire an ad bidding function before refreshing the ad with `DFPManager.refresh([options.slotId])`.

The original purpose of this PR is to get react-dfp to integrate with ad bidding systems such as Amazon's Transparent Ad Marketplace via APS [[link](https://aps.amazon.com/aps/transparent-ad-marketplace/)].

### CODE REVIEW

File: js/adslot.js [hunk 1]
```diff
@@ -21,10 +21,12 @@ export class AdSlot extends React.Component {
     adSenseAttributes: PropTypes.object,
     targetingArguments: PropTypes.object,
     onSlotRender: PropTypes.func,
+    onSlotRegister: PropTypes.func,
     onSlotIsViewable: PropTypes.func,
     onSlotVisibilityChanged: PropTypes.func,
     shouldRefresh: PropTypes.func,
     slotId: PropTypes.string,
+    wrapperClass: PropTypes.string,
```
Comment (jaanauati) [inline js/adslot.js]: this seems really useful!, can we rename it to 'extrassClasses'?

File: js/adslot.js [hunk 2]
```diff
@@ -127,6 +135,16 @@ export class AdSlot extends React.Component {
     }
   }
 
+  slotRegisterCallback() {
```
Comment (jaanauati) [inline js/adslot.js]: I love this!, the past months i've been thinking in a way to let the app code know what the slot-ids of the registered ads are, this is such a great solution.

File: js/dfpslotsprovider.js [hunk 3]
```diff
@@ -214,7 +219,7 @@ export default class DFPSlotsProvider extends React.Component {
   }
 
   render() {
-    const children = <div> {this.props.children} </div>;
+    const children = <React.Fragment> {this.props.children} </React.Fragment>;
```
Comment (jaanauati) [inline js/dfpslotsprovider.js]: I really like `<Fragment>`, but this component was introduced in ~16.X, react-dfp on the other hand is still supporting older versions like react 15.0 and 14.0 so this will not work, could u please revert this change?

**Comment by scott-thrillist** [comment]
@jaanauati code review comments addressed. Can you review? Hoping to get this merged and on NPM before our new site launches :]

File: js/dfpslotsprovider.js [hunk 4]
```diff
@@ -214,7 +219,11 @@ export default class DFPSlotsProvider extends React.Component {
   }
 
   render() {
-    const children = <div> {this.props.children} </div>;
+    let children = <div> {this.props.children} </div>;
+    if (React.Fragment) {
```
Comment (jaanauati) [inline js/dfpslotsprovider.js]: @scott-thrillist I'm really sorry to bother you about this again :-(, I really-really appreciate your contribution and I don't want to be a pita)  I'm trying to see react-dfp behavior from the point of the client apps, having react-dfp outputting different dom structure/behavior depending of the react version seems a bit confusing to me (and error prone, think in scenarios like debugging, styling, etc), don't you agree?

Do you really need this?, if that's the case i would really love to know why :-)

Comment (mmaxacov) [reply js/dfpslotsprovider.js]: IDK about OP, but for me, it would be great to have fragments there so that SSR hydration does not break if the context is not present. 

ATM I am injecting the context into the SSR jsx, but I'd rather not do it. So it's not a big issue per se.

WDYT about adding an optional parameter called something like `childrenWrapper` that defaults to Fragment? This would allow the legacy users to pass div or any component? Or, in order to not create a new major version, make it vice versa and allow passing the fragment with div being the default?

File: js/adslot.js [hunk 5]
```diff
@@ -21,10 +21,12 @@ export class AdSlot extends React.Component {
     adSenseAttributes: PropTypes.object,
     targetingArguments: PropTypes.object,
     onSlotRender: PropTypes.func,
+    onSlotRegister: PropTypes.func,
     onSlotIsViewable: PropTypes.func,
     onSlotVisibilityChanged: PropTypes.func,
     shouldRefresh: PropTypes.func,
     slotId: PropTypes.string,
+    extraClasses: PropTypes.string,
```
Comment (mmaxacov) [inline js/adslot.js]: WDYT about sticking to className (would allow styled-components and, I think, emotion, to hook in)?

File: js/dfpslotsprovider.js (hunk 4 above)
Comment (jaanauati) [reply js/dfpslotsprovider.js]: > IDK about OP, but for me, it would be great to have fragments there so that SSR hydration does not break if the context is not present.

As far I am aware, react-dfp should support SSR without issues, or at least it used to be the case... I wonder if this is a new bug introduced with the last updates I pushed to support the new context api.

@mmaxacov, could u please share a code snippet showing how your code looks? and also please tell me what version of react you are using.... i'll try to reproduce the bug at my side.

thanks

Comment (mmaxacov) [reply js/dfpslotsprovider.js]: @jaanauati we have a separate file for SSR that does not wrap the app with the provider in order to minimize the server execution time. It builds the HTML and sends in to the client. On the client, the app is being rendered again with HTML, but now there is the provider present. Since the provider wraps the whole app in a div, the hidration fails, since there is now one extra div at the top of the hierarchy.

It's not a big issue in general, so feel free to ignore it, but in my personal opinion, providers are an ethereal, logical abstraction and them rendering HTML looks spooky.

Comment (jaanauati) [reply js/dfpslotsprovider.js]: @mmaxacov 
> It's not a big issue in general, so feel free to ignore it, but in my personal opinion, providers are an ethereal, logical abstraction and them rendering HTML looks spooky.


I see your point now, thank you for the explanation!, I totally agree w you on this.

According @apascal the `<div>` wrapper may not be needed at all: 
https://github.com/jaanauati/react-dfp/pull/164

@scott-thrillist could u please revert this block :-) ? the discussion will continue in the other pull request. 

thanks!

File: js/adslot.js (hunk 5 above)
Comment (scott-thrillist) [reply js/adslot.js]: Good call, updated

File: js/dfpslotsprovider.js (hunk 4 above)
Comment (scott-thrillist) [reply js/dfpslotsprovider.js]: This is done

**Comment by jaanauati** [comment]
LGTM!, 
given that we are adding a new feature I think we should set version=0.15.0, don´t you agree?
